### PR TITLE
Refactor equipotentials

### PR DESCRIPTION
**Pull request by eggrobin**
No functional change.

This should make it possible to tweak the logic that chooses the energy levels, and to do some much-needed benchmarking.

### CODE REVIEW

File: physics/equipotential_test.cpp [hunk 1]
```diff
@@ -368,153 +370,52 @@ TEST_F(EquipotentialTest, DISABLED_RotatingPulsating_GlobalOptimization) {
...
-           {earth_position, earth->min_radius() / r(t) * (1 * Metre)}},
-          [](Position<World> q) {
-            return World::origin + Normalize(q - World::origin) * 3 * Metre;
-          },
-          energy);
       for (auto const& line : lines) {
         std::vector<Position<World>>& equipotential =
             equipotentials_at_energy.emplace_back();
-        for (auto const& [s, dof] : line) {
+        for (auto const& [t, dof] : line) {
```
Comment (pleroy) [inline physics/equipotential_test.cpp:410]: Make this `_`, you don't use it.

File: physics/lagrange_equipotentials_body.hpp [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+  // The distance between the bodies is 1 m in the rotating-pulsating frame, and
+  // the origin is at their barycentre.  A 3 m box is large enough to encompass
+  // the L₄ and L₅ Lagrange points, as well as the whole ridge they sit on,
+  // regardless of the mass ratio.
+  const MultiLevelSingleLinkage<SpecificEnergy,
+                                Position<RotatingPulsating>,
+                                2>::Box box = {
+      .centre = RotatingPulsating::origin,
+      .vertices = {
+          Displacement<RotatingPulsating>({3 * Metre, 0 * Metre, 0 * Metre}),
```
Comment (pleroy) [inline physics/lagrange_equipotentials_body.hpp:71]: Let's gather all the constants (like `3 * Metre`) at the beginning of the body.  Some of them we might want to tune or move to the blueprint.

File: physics/lagrange_equipotentials_body.hpp [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+          Displacement<RotatingPulsating>({0 * Metre, 3 * Metre, 0 * Metre})}};
+
+  // This is a length in the rotating-pulsating frame, so really one billionth
+  // of the distance between the bodies.
+  constexpr Length characteristic_length = 1 * Nano(Metre);
```
Comment (pleroy) [inline physics/lagrange_equipotentials_body.hpp:76]: Move the constant.

File: physics/lagrange_equipotentials_body.hpp [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+  Equipotential<Inertial, RotatingPulsating> const equipotential(
+      {EmbeddedExplicitRungeKuttaIntegrator<
+           DormandPrince1986RK547FC,
+           Equipotential<Inertial, RotatingPulsating>::ODE>(),
+       /*max_steps=*/1000,
```
Comment (pleroy) [inline physics/lagrange_equipotentials_body.hpp:81]: Move the constant.

File: physics/lagrange_equipotentials_body.hpp [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+       /*length_integration_tolerance=*/characteristic_length},
+      &reference_frame,
+      characteristic_length);
+
+  // TODO(phl): Make this interruptible.
+  auto const arg_maximorum =
+      MultiLevelSingleLinkage<SpecificEnergy, Position<RotatingPulsating>, 2>(
+          box, potential, gradient)
+          .FindGlobalMaxima(
+              /*points_per_round=*/1000,
```
Comment (pleroy) [inline physics/lagrange_equipotentials_body.hpp:91]: Move the constant.

File: physics/lagrange_equipotentials_body.hpp [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+              /*number_of_rounds=*/std::nullopt,
+              /*local_search_tolerance=*/1e-3 * Metre);
```
Comment (pleroy) [inline physics/lagrange_equipotentials_body.hpp:93]: Move the constant.

File: physics/lagrange_equipotentials_body.hpp [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+  SpecificEnergy maximum_maximorum = -Infinity<SpecificEnergy>;
+  for (auto const& arg_maximum : arg_maximorum) {
+    auto const maximum = potential(arg_maximum);
+    if (!IsFinite(maximum)) {
+      return absl::OutOfRangeError(absl::StrCat("Improper maximum ",
```
Comment (pleroy) [inline physics/lagrange_equipotentials_body.hpp:98]: "Improper" is weird, "singular"?

File: physics/lagrange_equipotentials_body.hpp [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+          primary_Inertial_position.Get());
+  Position<RotatingPulsating> const secondary_position =
+      reference_frame.ToThisFrameAtTimeSimilarly(t).similarity()(
+          secondary_Inertial_position.Get());
+
+  Length primary_radius;
+  if (parameters.primaries.size() == 1) {
+    primary_radius = parameters.primaries.front()->min_radius();
+  } else {
+    for (not_null const primary : parameters.primaries) {
```
Comment (pleroy) [inline physics/lagrange_equipotentials_body.hpp:136]: Could we document this computation?  Each time I read it I need to think.

File: physics/lagrange_equipotentials_body.hpp [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+                       secondary->min_radius());
+    }
+  }
+
+  // TODO(egg): Consider additional wells.
+  std::vector<Equipotential<Inertial, RotatingPulsating>::Well> wells{
+      {secondary_position, secondary_radius / r * (1 * Metre)},
+      {primary_position, primary_radius / r * (1 * Metre)}};
+
+  double const arg_approx_l1 = Brent(
```
Comment (pleroy) [inline physics/lagrange_equipotentials_body.hpp:165]: Comment this computation similarly to lines 176-179.

File: physics/lagrange_equipotentials_body.hpp [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+      [&](double const x) {
+        return potential(
+            Barycentre(std::pair(secondary_position, primary_position),
+                       std::pair(x, 1 - x)));
+      },
```
Comment (pleroy) [inline physics/lagrange_equipotentials_body.hpp:170]: Give a name to this lambda and use it at line 189.

File: physics/lagrange_equipotentials_body.hpp [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+      0.0,
+      1.0,
+      std::greater<>{});
+  double const arg_approx_l2 = Brent(
+      [&](double x) {
```
Comment (pleroy) [inline physics/lagrange_equipotentials_body.hpp:175]: Same.

File: physics/lagrange_equipotentials_body.hpp [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+  // equipotential usually distinguishes a small region around L₄/L₅ from the
+  // rest of the ridge that contains them and L₃.
+  SpecificEnergy const l45_separator =
+      maximum_maximorum - (maximum_maximorum - approx_l1_energy) /
+                              (4 * Sqrt(primary_Inertial_position.weight() /
+                                        secondary_Inertial_position.weight()));
+
+  auto const towards_infinity = [](Position<RotatingPulsating> q) {
+    return RotatingPulsating::origin +
+           Normalize(q - RotatingPulsating::origin) * 3 * Metre;
```
Comment (pleroy) [inline physics/lagrange_equipotentials_body.hpp:209]: Move the constant.

File: physics/lagrange_equipotentials_body.hpp [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+                    ephemeris_->trajectory(secondary)->EvaluatePosition(t))
+                           .Norm() +
+                       secondary->min_radius());
+    }
+  }
+
+  // TODO(egg): Consider additional wells.
+  std::vector<Equipotential<Inertial, RotatingPulsating>::Well> wells{
+      {secondary_position, secondary_radius / r * (1 * Metre)},
+      {primary_position, primary_radius / r * (1 * Metre)}};
```
Comment (pleroy) [inline physics/lagrange_equipotentials_body.hpp:163]: Move the constants.

Comment (eggrobin) [reply physics/lagrange_equipotentials_body.hpp:163]: Not this one: it is not something we want to parameterize, it is instead from the contract of RotatingPulsatingFrame.

**Pull request merged**